Change hspec-discover sort from lexicographic to to natural order

ephemient opened the pull request:
I would like my tests to be [sorted for humans](https://blog.codinghorror.com/sorting-for-humans-natural-sort-order/). It seems easy to add to `hspec-discover`, would you consider something like this?

sol commented:
@ephemient Hey, thanks for taking the time to open this PR!

> I would like my tests to be sorted for humans. It seems easy to add to hspec-discover, would you consider something like this?

From my perspective we can consider to make this the default behavior.  However, `natural-sort` has transitive dependencies on `text` and `parsec` which I would prefer to prevent.

What we would need is a standalone implementation of natural sorting that we vendor with Hspec.

sol commented:
One more thing, contributions to Hspec require tests prior to being considered ;)

ephemient commented:
@sol I've updated the PR to drop the dependency, replacing it with a home-grown natural sort, and added a `Spec` for it. Well, just the sorting function; I haven't dug into how to add it to `RunSpec` or the integration tests yet.

sol commented on hspec-discover/src/Test/Hspec/Discover/Sort.hs:
```diff
@@ -0,0 +1,38 @@
+(~>) :: Ordering -> Ordering -> Ordering
+EQ ~> a = a
+a ~> _ = a
+infixr 5 ~>
```
- I think this is the same as `mappend` / `<>`.
- Even if we were to implement this ourself, I don't see the benefit of defining a operator for one-way use.  What I would want instead is a name that conveys what this function does.

ephemient replied:
Oh I didn't realize there was a Monoid instance for Ordering. That's exactly what this is, great.

sol commented on hspec-discover/src/Test/Hspec/Discover/Sort.hs:
```diff
@@ -0,0 +1,38 @@
+compareLength [] [] = EQ
+compareLength [] _ = LT
+compareLength _ [] = GT
+compareLength (_:xs) (_:ys) = compareLength xs ys
```
This can be replaced with

```haskell
compareLength :: [a] -> [a] -> Ordering
compareLength = comparing length
```

sol commented on hspec-discover/test/Test/Hspec/Discover/RunSpec.hs:
```diff
@@ -38,7 +41,11 @@ spec = do
+          , ">> postProcessSpec \"test-data/nested-spec/Dir8/X128Spec.hs\" (describe \"Dir8.X128\" Dir8.X128Spec.spec)"
+          , ">> postProcessSpec \"test-data/nested-spec/Dir16/X48Spec.hs\" (describe \"Dir16.X48\" Dir16.X48Spec.spec)"
+          , ">> postProcessSpec \"test-data/nested-spec/Dir16/X160Spec.hs\" (describe \"Dir16.X160\" Dir16.X160Spec.spec)"
+          , ">> postProcessSpec \"test-data/nested-spec/Foo/Bar/BazSpec.hs\" (describe \"Foo.Bar.Baz\" Foo.Bar.BazSpec.spec)"
```
I think there is no really need to test this here.  Unit tests for the sort function are good enough.

sol commented on hspec-discover/test/Test/Hspec/Discover/SortSpec.hs:
```diff
@@ -0,0 +1,49 @@
+      it "numeric segments" $ do
+        "3.1.415" <=> "3.14.15" `shouldBe` LT
+        "3.14.15" <=> "3.1.415" `shouldBe` GT
+      it "numeric ties by string" $ do
```
I think this is the right thing to do :clap:

Many other implementations for natural sorting miss this case and as a result assume that e.g. `00` and `0` are equal.

ephemient replied:
It was only after writing this test that I realized I wanted to add it to the comparator :)

sol commented on hspec-discover/test/Test/Hspec/Discover/SortSpec.hs:
```diff
@@ -0,0 +1,49 @@
+          ]
+  where
+    (<=>) = comparing naturalSortKey
+    infix 4 <=>
```
Not too impressed by this spec.  May I ask you if you developed the code test-driven?  Or did the tests come as an afterthought?

ephemient replied:
Satisfying the `z(\d+).txt` image in the blog post I linked to was goal as I was writing this, although it was tested in GHCi without this spec at first, then I added the ones above as they were also things I was checking interactively.

sol replied:
Don't test interactively, use [`hspec/sensei`](https://github.com/hspec/sensei) ;)

sol commented on hspec-discover/test/Test/Hspec/Discover/SortSpec.hs:
```diff
@@ -0,0 +1,49 @@
+        "Hello2World!0" <=> "Hello2World9" `shouldBe` GT
+      it "numeric segments" $ do
+        "3.1.415" <=> "3.14.15" `shouldBe` LT
+        "3.14.15" <=> "3.1.415" `shouldBe` GT
```
You may be able to make some of the things you test here redundant by deriving the `Ord` instance, e.g.:

```haskell
import           Data.Char

data NaturalSortChunk = NumericChunk Integer Int | StringChunk String
  deriving (Eq, Show, Ord)

naturalSortKey :: String -> [NaturalSortChunk]
naturalSortKey s@(c : _)
  | isDigit c = NumericChunk (read num) (length num) : naturalSortKey afterNum
  | otherwise = StringChunk str : naturalSortKey afterStr
  where
    (num, afterNum) = span  isDigit s
    (str, afterStr) = break isDigit  s
naturalSortKey [] = []
```

ephemient replied:
Ah sure, I suppose since `isDigit` is ASCII digits only it's not like there can be ties that need to be broken by string.

sol replied:
The code I give still uses the string length as a tie breaker.  This is why is `NumericChunk Integer Int`.  The `Int` is the length of the string.

sol replied:
And yes, there are ties, e.g. `01` and `1`, see below.

ephemient replied:
I meant, broken by string comparison in addition to by length. That could happen with Unicode digits but (numeric value, length) should be sufficient with ASCII only.

sol commented on hspec-discover/src/Test/Hspec/Discover/Sort.hs:
```diff
@@ -0,0 +1,38 @@
+naturalSortKey :: String -> [NaturalSortChunk]
+naturalSortKey s@(c:_)
+  | isDigit c = NumericChunk (read num) num : naturalSortKey afterNum
+  | otherwise = StringChunk             str : naturalSortKey afterStr
```
Aligning code, while it can look nice, has it's downsides.  Here, if you change the `isDigit c` case, then you may also need to change the `otherwise` case to keep things aligned, which gives you a bigger diff, increases the chance for merge conflicts  and puts more burden on the code reviewer.  This is why I try to avoid aligning out of pragmatism.

sol commented:
If we already tackle this, we may also want to sort alphabetically, e.g `A a B b` instead of `A B a b`.

sol commented on hspec-discover/test/Test/Hspec/Discover/SortSpec.hs:
```diff
@@ -0,0 +1,49 @@
+        "" <=> "" `shouldBe` EQ
+        "Hello2World" <=> "Hello2World" `shouldBe` EQ
+        "123Hello456" <=> "123Hello456" `shouldBe` EQ
+        "" <=> "" `shouldBe` EQ
```
Provided you derive the instance, I don't think you will need this test.  But if it were that you really want to test this, then this would be a prime use case for QuickCheck, e.g.:

```haskell
it "is reflexive" $ do
  property $ \ x -> x <=> x `shouldBe` EQ
```

ephemient commented:
I've been distracted by [AoC](http://adventofcode.com/) and the holidays, but I'll be back to continue cleaning up this PR. Thanks for looking.

> If we already tackle this, we may also want to sort alphabetically, e.g A a B b instead of A B a b.

I thought of that, but held off because there's a tarpit of potential locale-based issues that I'd defer to Text on.

On the other hand, since these are module names maybe there's no such issues.

sol commented:
> I thought of that, but held off because there's a tarpit of potential locale-based issues that I'd defer to Text on.

I assume changing to `StringChunk String String` and then doing `StringChunk (map toLower str) str` will do the trick with a derived `Ord` instance.

sol commented:
> I've been distracted by AoC and the holidays

No worries, no need to rush ;)

ephemient commented:
> I assume changing to `StringChunk String String` and then doing `StringChunk (map toLower str) str` will do the trick with a derived `Ord` instance.

Not quite the behavior I would like - I changed it to postpone the original string comparison until after all the chunks are compared.

sol commented on hspec-discover/test/Test/Hspec/Discover/SortSpec.hs:
```diff
@@ -0,0 +1,57 @@
+      it "symmetrically" $ property $ \a b -> do
+        a <=> b === compare EQ (b <=> a)
+      it "transitively" $ property $ \a b c -> do
+        a <=> b == b <=> c ==> a <=> b === a <=> c
```
I think eflexivity, symmetry and transitivity is always given for derived instances, so there is no need to test for this.

We might want to test for injectivity though, see https://github.com/hspec/hspec/commit/62e0c4437c7fb88179491941081257b9cdbb8096.

ephemient replied:
Good point. Thanks for merging and cleaning up 😃

sol commented on hspec-discover/test/Test/Hspec/Discover/SortSpec.hs:
```diff
@@ -0,0 +1,57 @@
+        "HelloWorld2" <=> "helloworld1" `shouldBe` GT
+      it "string ties by case" $ do
+        "HelloWorld3" <=> "Helloworld3" `shouldBe` LT
+        "Helloworld3" <=> "HelloWorld3" `shouldBe` GT
```
https://github.com/hspec/hspec/commit/11837f69e345092897d3b360164f6b2f07747225

I think for derived `Ord` instances `a < b <=> b > a` is always true.

sol commented:
@ephemient looking at the following example:

```
SpecFoo.hs
SpecFoo.lhs
Specfoo.hs
Specfoo.lhs
```

we want that order.  We don't want the spec ordering depend on the file extension.   But the current implementation would give us:

```
SpecFoo.hs
Specfoo.hs
SpecFoo.lhs
Specfoo.lhs
```

I changed this, see https://github.com/hspec/hspec/commit/65b246ef0840e5b3f4cf3567c741d1edc33debb8.

ephemient commented:
Ah hmm, so the reason I did it differently is because I didn't like `FooBar2 < Foobar1`. But I agree that having modules sorted by extension is worse.
I suppose another workaround could be to compare without the extension first. But that seems more complex than necessary.
